beacon-chain/casper/incentives.go

```diff
@@ -1,37 +1,102 @@
 package casper
 
 import (
+	"math"
+
 	"github.com/prysmaticlabs/prysm/beacon-chain/params"
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
-	"github.com/prysmaticlabs/prysm/shared"
 	"github.com/sirupsen/logrus"
 )
 
 var log = logrus.WithField("prefix", "casper")
 
 // CalculateRewards adjusts validators balances by applying rewards or penalties
 // based on FFG incentive structure.
+// FFG Rewards scheme rewards validator who have voted on blocks, and penalises those validators
+// who are offline. The penalties are more severe the longer they are offline.
 func CalculateRewards(
-	attestations []*pb.AggregatedAttestation,
+	slot uint64,
+	voterIndices []uint32,
 	validators []*pb.ValidatorRecord,
 	dynasty uint64,
-	totalDeposit uint64,
-	totalParticipatedDeposit uint64) ([]*pb.ValidatorRecord, error) {
+	totalParticipatedDeposit uint64,
+	timeSinceFinality uint64) ([]*pb.ValidatorRecord, error) {
+	totalDeposit := TotalActiveValidatorDeposit(dynasty, validators)
 	activeValidators := ActiveValidatorIndices(validators, dynasty)
-	attesterBitfield := attestations[len(attestations)-1].AttesterBitfield
-	attesterFactor := totalParticipatedDeposit * 3
-	totalFactor := totalDeposit * 2
-
-	if attesterFactor >= totalFactor {
-		log.Debug("Applying rewards and penalties for the validators from last cycle")
-		for i, attesterIndex := range activeValidators {
-			voted := shared.CheckBit(attesterBitfield, int(attesterIndex))
-			if voted {
-				validators[i].Balance += params.AttesterReward
-			} else {
-				validators[i].Balance -= params.AttesterReward
+	rewardQuotient := uint64(RewardQuotient(dynasty, validators))
+	PenaltyQuotient := uint64(QuadraticPenaltyQuotient())
```

Comment:
should `PenaltyQuotient` be exported?

Reply:
I think I will leave all the incentives unexported then since, they are only used in the Casper Package

Reply:
That sounds good : )
also why is `PenaltyQuotient` capitalized? 😅

Reply:
Haha , I have no idea, I just changed it